pkg/kapis/devops/v1alpha4/handler.go

```
@@ -0,0 +1,116 @@
+	return func(request *restful.Request, response *restful.Response) {
+		nsName := request.PathParameter("namespace")
+		pipName := request.PathParameter("pipeline")
+		branchName := request.QueryParameter("branch")
+		backward, err := strconv.ParseBool(request.QueryParameter("backward"))
+		if err != nil {
+			// by default, we have to guarantee backward compatibility
+			backward = true
+		}
+
+		queryParam := query.ParseQueryParameter(request)
+
+		// validate the Pipeline
+		var pip v1alpha3.Pipeline
+		err = h.client.Get(context.Background(), client.ObjectKey{Namespace: nsName, Name: pipName}, &pip)
+		if err != nil {
+			api.HandleError(request, response, err)
+			return
+		}
+
+		// build label selector
+		labelSelector, err := buildLabelSelector(queryParam, pip.Name, branchName)
+		if err != nil {
+			api.HandleError(request, response, err)
+			return
+		}
+
+		var prs v1alpha4.PipelineRunList
+		// fetch PipelineRuns
+		if err := h.client.List(context.Background(), &prs,
+			client.InNamespace(pip.Namespace),
+			client.MatchingLabelsSelector{Selector: labelSelector}); err != nil {
+			api.HandleError(request, response, err)
+			return
+		}
+		var transformFunc = v1alpha32.NoTransformFunc()
+		if backward {
+			transformFunc = backwardTransform()
+		}
+
+		apiResult := v1alpha32.DefaultList(convertPipelineRunsToObject(prs.Items), queryParam, v1alpha32.DefaultCompare(), v1alpha32.DefaultFilter(), transformFunc)
+		_ = response.WriteAsJson(apiResult)
+	}
+}
+
+// backwardTransform transforms PipelineRun into JSON raw message of Jenkins run status.
+func backwardTransform() v1alpha32.TransformFunc {
+	return func(object runtime.Object) interface{} {
+		pr := object.(*v1alpha4.PipelineRun)
+		runStatusJSON := pr.Annotations[v1alpha4.JenkinsPipelineRunStatusKey]
+		rawRunStatus := json.RawMessage{}
+		// the error will never happen due to raw message won't be nil, so we ignore the error explicitly
+		_ = rawRunStatus.UnmarshalJSON([]byte(runStatusJSON))
```

what will happen if the `runStatusJSON`was modified manually, and it turns out to an invalid json string.

---

How about returning `{}` if the JSON is invalid?

---

No matter what, cannot eat this error.

---

Please see the first element below:

```json
{
 "items": [
  {},
  {
   "causes": [
    {
     "_class": "io.jenkins.blueocean.service.embedded.rest.AbstractRunImpl$BlueCauseImpl",
     "shortDescription": "Started by user admin",
     "userId": "admin",
     "userName": "admin"
    }
   ],
   "durationInMillis": 62234,
   "enQueueTime": "2021-09-08T21:57:55.633+08:00",
   "endTime": "2021-09-08T21:58:57.876+08:00",
   "estimatedDurationInMillis": 57378,
   "id": "5",
   "organization": "jenkins",
   "pipeline": "main",
   "replayable": true,
   "result": "SUCCESS",
   "runSummary": "stable",
   "startTime": "2021-09-08T21:57:55.642+08:00",
   "state": "FINISHED",
   "type": "WorkflowRun"
  },
  {
   "causes": [
    {
     "_class": "io.jenkins.blueocean.service.embedded.rest.AbstractRunImpl$BlueCauseImpl",
     "shortDescription": "Started by user admin",
     "userId": "admin",
     "userName": "admin"
    }
   ],
   "durationInMillis": 64198,
   "enQueueTime": "2021-09-08T14:22:14.762+08:00",
   "endTime": "2021-09-08T14:23:18.969+08:00",
   "estimatedDurationInMillis": 62034,
   "id": "4",
   "organization": "jenkins",
   "pipeline": "main",
   "replayable": true,
   "result": "SUCCESS",
   "runSummary": "stable",
   "startTime": "2021-09-08T14:22:14.771+08:00",
   "state": "FINISHED",
   "type": "WorkflowRun"
  }
 ],
 "totalItems": 3
}
```

---

It's not reasonable. There are 3 items. But one of them is empty.

---

Not necessary to return the message of invalid data to API response. Please use a logger to print it.

---

Consider that there is no Jenkins run status yet, what data should be returned

---

I prefer to return nothing. In this case, the total number should be 2.